pyemma/plots/plots2d.py

```diff
@@ -156,33 +153,403 @@ def plot_free_energy(xall, yall, weights=None, ax=None, nbins=100, ncountours=10
+            vmax = _np.max(z)
+    cs = ax.contourf(
+        x, y, z, ncontours, norm=norm,
+        vmin=vmin, vmax=vmax, cmap=cmap,
+        levels=levels)
     if cbar:
-        cbar = _plt.colorbar(CS)
+        if cax is None:
+            cbar = fig.colorbar(cs, ax=ax)
+        else:
+            cbar = fig.colorbar(cs, cax=cax)
         if cbar_label is not None:
-            cbar.ax.set_ylabel(cbar_label)
+            cbar.set_label(cbar_label)
+    else:
+        cbar = None
+    return fig, ax, cbar
+
+
+# ######################################################################
+#
+# new plotting functions
+#
+# ######################################################################
+    
+
+def plot_density(
+        xall, yall, weights=None, ax=None, cmap='Blues',
+        nbins=100, ncontours=100, avoid_zero_count=False,
+        cbar=True, cax=None, cbar_label='density',
+        logscale=False):
+    """Plot a two-dimensional density map.
+
+    Parameters
+    ----------
+    xall : ndarray(T)
+        Sample x-coordinates.
+    yall : ndarray(T)
+        Sample y-coordinates.
+    weights : ndarray(T), default=None
+        Sample weights; by default all samples have the same weight.
+    ax : matplotlib.Axes object, optional, default=None
+        The ax to plot to; if ax=None, a new ax (and fig) is created.
+        Number of contour levels.
+    cmap : matplotlib colormap, optional, default='Blues'
+        The color map to use.
+    nbins : int, default=100
+        Number of histogram bins used in each dimension.
+    ncontours : int, optional, default=100
+    avoid_zero_count : bool, default=False
+        Avoid zero counts by lifting all histogram elements to the
+        minimum value before computing the free energy. If False,
+        zero histogram counts would yield infinity in the free energy.
+    vmin : float, optional, default=None
+    cbar : boolean, optional, default=True
+        Plot a color bar.
+    cax : matplotlib.Axes object, optional, default=None
+        Plot the colorbar into a custom axes object instead of
+        stealing space from ax.
+    cbar_label : str, optional, default='density'
+        Colorbar label string; use None to suppress it.
+    logscale : boolean, optional, default=False
+        Plot the density values in logscale.
+
+    Returns
+    -------
+    fig : matplotlib.Figure object
+        The figure in which the used ax resides.
+    ax : matplotlib.Axes object
+        The ax in which the map was plotted.
+    cbar : matplotlib.Colorbar object
+        The corresponding colorbar object; None if no colorbar
+        was requested.
+
+    """
+    x, y, z = get_histogram(
+        xall, yall, nbins=nbins,weights=weights,
+        avoid_zero_count=avoid_zero_count)
+    return plot_map(
+        x, y, _to_density(z).T, ncontours=ncontours, ax=ax,
+        cmap=cmap, cbar=cbar, cax=cax, cbar_label=cbar_label,
+        logscale=logscale)
+    
+
+def plot_free_energy_new(
```

Comment:
what is actually the difference to the old method? Can we decide to have only one or are they fundamentally different?
A short line in the docstring what this method does different would be nice.

Reply:
The new plotting functions always return `fig, ax, cbar` while the old functions return either `fig, ax` or just `ax`. Further, `plot_free_energy` has a parameter typo (`countour`) which has been fixed in the new function.

I would love to immediately replace everything with the new functions but this would slightly break the API.

Reply:
We could have a legacy switch parameter to steer the old or new behaviour, but this would convolute the code a bit. I think we should deprecate the old function and promote the new one, shouldn't we?

Reply:
Yes, indeed. I don't think functions would be too convoluted, though. Using a legacy switch is a good idea!

Reply:
Do we need to conserve the sequence of named parameters in legacy mode?